library/system/test/ui_plugin_info_test.rb

```diff
@@ -0,0 +1,306 @@
+        expect(subject.ui_so_major).to eq "15"
+      end
+    end
+
+    context "with the Qt UI + Qt-Pkg extension" do
+      let(:maps_file) { stored_proc_maps("qt-pkg") }
+
+      it "identifies the UI SO major number as 15" do
+        expect(subject.ui_so_major).to eq "15"
+      end
+    end
+
+    context "without any UI" do
+      let(:maps_file) { stored_proc_maps("no-ui") }
+
+      it "returns nil" do
+        expect(subject.ui_so_major).to eq nil
+      end
+    end
+  end
+
+  describe "#ui_extension_plugin" do
+    let(:subject) { described_class.new(maps_file) }
+
+    context "empty" do
+      let(:maps_file) { nil }
+
+      it "does not crash and burn" do
+        expect(subject.ui_extension_plugin("pkg")).to eq nil
+      end
+    end
+
+    context "with the Qt UI" do
+      let(:maps_file) { stored_proc_maps("qt") }
+
+      it "returns the -pkg counterpart" do
+        expect(subject.ui_extension_plugin("pkg")).to eq "libyui-qt-pkg.so.15.0.0"
+      end
+    end
+
+    context "with the Qt UI + Qt-Graph extension" do
+      let(:maps_file) { stored_proc_maps("qt-graph") }
+
+      it "returns the -pkg counterpart" do
+        expect(subject.ui_extension_plugin("pkg")).to eq "libyui-qt-pkg.so.15.0.0"
+      end
+    end
+
+    context "with the NCurses UI" do
+      let(:maps_file) { stored_proc_maps("ncurses") }
+
+      it "returns the -pkg counterpart" do
+        expect(subject.ui_extension_plugin("pkg")).to eq "libyui-ncurses-pkg.so.15.0.0"
+      end
+    end
+
+    context "without any UI" do
+      let(:maps_file) { stored_proc_maps("no-ui") }
+
+      it "returns nil" do
+        expect(subject.ui_extension_plugin("pkg")).to eq nil
+      end
+    end
+  end
+
+  describe "#ui_extension_plugin_complete" do
+    let(:subject) { described_class.new(maps_file) }
+
+    context "empty" do
+      let(:maps_file) { nil }
+
+      it "does not crash and burn" do
+        expect(subject.ui_extension_plugin_complete("pkg")).to eq nil
+      end
+    end
+
+    context "with the Qt UI" do
+      let(:maps_file) { stored_proc_maps("qt") }
+
+      it "returns the -pkg counterpart" do
+        expect(subject.ui_extension_plugin_complete("pkg")).to eq "/usr/lib64/yui/libyui-qt-pkg.so.15.0.0"
+      end
+    end
+
+    context "with the NCurses UI" do
+      let(:maps_file) { stored_proc_maps("ncurses") }
+
+      it "returns the -pkg counterpart" do
+        expect(subject.ui_extension_plugin_complete("pkg")).to eq "/usr/lib64/yui/libyui-ncurses-pkg.so.15.0.0"
+      end
+    end
+
+    context "without any UI" do
+      let(:maps_file) { stored_proc_maps("no-ui") }
+
+      it "returns nil" do
+        expect(subject.ui_extension_plugin_complete("pkg")).to eq nil
+      end
+    end
+  end
+
+  describe "#ui_extension_pkg" do
+    let(:subject) { described_class.new(maps_file) }
+
+    context "empty" do
+      let(:maps_file) { nil }
+
+      it "does not crash and burn" do
```

Comment:
This is repeated several times in the tests. See the RSpec [shared examples](https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples) documentation.

Reply:
Maybe next time. I want to get finished with this; it already took a lot longer than expected.

I find the RSpec documentation largely useless; it's the thing that costs me most time and life energy (by far!). Dealing with that test framework always involves a lot of frustration, looking up existing examples in our other code because the official documentation is so completely worthless, a lot of screaming and a strong urge to strangle everbody involved with that stuff.

They are just too damn lazy to write any good documentation, and they keep getting away with it.

And the completely useless error messages that thing throws when it doesn't like something are another dimension of un-usability; they might as well only report "something is wrong" (because that's what it comes down to).

Having said this, reading existing RSpec tests which they claim is so self-explanatory always involves a lot of reverse engineering their completely crazy syntax to make sense of anything. This is a tool on the verge (and oftentimes beyond the verge) of being unusable.

Why we put up with this level of crapware is completely beyond me.

Reply:
OK, we can refactor the tests later...